### PR TITLE
Adding CVE/CWE utilities

### PR DESCRIPTION
**Pull request by MarcJose**

### CODE REVIEW

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 68
           

Complexity increasing per file
==============================
- varats/utils/security_util.py  8
- varats/tools/driver_security.py  3
- varats/security.py  5
- tests/utils/test_security.py  3
- varats/tools/security_map.py  14
         

Clones added
============
- varats/tools/security_map.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5174691&bid=16985828)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 22
           

Complexity increasing per file
==============================
- tests/utils/test_security.py  3
- varats/tools/driver_security.py  3
- varats/tools/security_map.py  14
- varats/utils/security_util.py  8
- varats/security.py  5
         

Clones added
============
- varats/tools/security_map.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5174691&bid=16985828)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 10
           

Complexity increasing per file
==============================
- varats/tools/driver_cve.py  3
- varats/data/provider/cve/cve.py  8
- varats/data/provider/provider.py  2
- varats/data/provider/cve/cve_provider.py  5
- varats/data/provider/cve/cve_map.py  14
- tests/data/test_cve.py  3
         

Clones added
============
- varats/data/provider/cve/cve_map.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5174691&bid=16985828)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 15
           

Complexity increasing per file
==============================
- varats/plots/cve_annotation.py  8
- varats/utils/git_util.py  1
- tests/data/test_cve.py  3
- varats/plots/blame_interaction_degree.py  2
- varats/data/provider/provider.py  2
- varats/data/provider/cve/cve_map.py  14
- varats/tools/driver_cve.py  3
- varats/data/provider/cve/cve_provider.py  5
- varats/data/provider/cve/cve.py  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5174691&bid=16985828)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 10
           

Complexity increasing per file
==============================
- varats/plots/cve_annotation.py  8
- varats/data/provider/cve/cve_map.py  14
- varats/plots/blame_interaction_degree.py  2
- varats/data/provider/cve/cve.py  8
- tests/data/test_cve.py  3
- varats/data/provider/cve/cve_provider.py  5
- varats/tools/driver_cve.py  3
- varats/utils/git_util.py  1
- varats/data/provider/provider.py  2
         

Clones added
============
- varats/projects/c_projects/lrzip.py  1
- varats/projects/c_projects/coreutils.py  1
- varats/projects/c_projects/opus.py  1
- varats/projects/c_projects/glibc.py  1
- varats/projects/c_projects/gzip.py  1
- varats/projects/c_projects/tmux.py  1
- varats/projects/c_projects/vim.py  1
- varats/projects/c_projects/libvpx.py  1
- varats/projects/c_projects/openssl.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5174691&bid=16985828)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 10
           

Complexity increasing per file
==============================
- varats/plots/cve_annotation.py  8
- varats/tools/driver_cve.py  3
- varats/data/provider/cve/cve_provider.py  5
- varats/plots/blame_interaction_degree.py  2
- tests/data/test_cve.py  3
- varats/data/provider/cve/cve_map.py  14
- varats/utils/git_util.py  1
- varats/data/provider/provider.py  2
- varats/data/provider/cve/cve.py  8
         

Clones added
============
- varats/projects/c_projects/coreutils.py  1
- varats/projects/c_projects/tmux.py  3
- varats/projects/c_projects/openssl.py  1
- varats/projects/c_projects/glibc.py  1
- varats/projects/c_projects/vim.py  1
- varats/projects/c_projects/git.py  1
- varats/projects/c_projects/lrzip.py  2
- varats/projects/c_projects/libvpx.py  1
- varats/projects/c_projects/gzip.py  1
- varats/projects/c_projects/opus.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5174691&bid=16985828)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- varats/plots/cve_annotation.py  8
- varats/data/provider/cve/cve_map.py  14
- varats/utils/git_util.py  1
- varats/data/provider/provider.py  2
- varats/plots/blame_interaction_degree.py  2
- tests/data/test_cve.py  3
- varats/data/provider/cve/cve_provider.py  5
- varats/data/provider/cve/cve.py  8
- varats/tools/driver_cve.py  3
         

Clones added
============
- varats/projects/c_projects/coreutils.py  1
- varats/projects/c_projects/vim.py  1
- varats/projects/c_projects/openssl.py  1
- varats/projects/c_projects/glibc.py  1
- varats/projects/c_projects/tmux.py  2
- varats/projects/c_projects/libvpx.py  1
- varats/projects/c_projects/opus.py  2
- varats/projects/c_projects/lrzip.py  2
- varats/projects/c_projects/gzip.py  1
- varats/projects/c_projects/git.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5174691&bid=16985828)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- varats/tools/driver_cve.py  3
- varats/utils/git_util.py  1
- varats/data/provider/cve/cve.py  8
- varats/data/provider/cve/cve_provider.py  5
- varats/data/provider/cve/cve_map.py  14
- varats/plots/blame_interaction_degree.py  2
- tests/data/test_cve.py  3
- varats/data/provider/provider.py  2
- varats/plots/cve_annotation.py  8
         

Clones added
============
- varats/projects/c_projects/coreutils.py  1
- varats/projects/c_projects/opus.py  2
- varats/projects/c_projects/git.py  1
- varats/projects/c_projects/gzip.py  1
- varats/projects/c_projects/lrzip.py  2
- varats/projects/c_projects/openssl.py  1
- varats/projects/c_projects/vim.py  1
- varats/projects/c_projects/libvpx.py  1
- varats/projects/c_projects/tmux.py  2
- varats/projects/c_projects/glibc.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5174691&bid=16985828)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Complexity increasing per file
==============================
- varats/plots/cve_annotation.py  8
- varats/tools/driver_cve.py  3
- varats/utils/git_util.py  1
- varats/plots/blame_interaction_degree.py  2
- varats/data/provider/cve/cve_map.py  14
- varats/data/provider/cve/cve.py  8
- varats/data/provider/cve/cve_provider.py  5
- varats/data/provider/provider.py  2
- tests/data/test_cve.py  3
         

Clones added
============
- varats/projects/c_projects/openssl.py  1
- varats/projects/c_projects/coreutils.py  1
- varats/projects/c_projects/vim.py  1
- varats/projects/c_projects/glibc.py  1
- varats/projects/c_projects/tmux.py  2
- varats/projects/c_projects/lrzip.py  2
- varats/projects/c_projects/git.py  1
- varats/projects/c_projects/opus.py  2
- varats/projects/c_projects/gzip.py  1
- varats/projects/c_projects/libvpx.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5174691&bid=16985828)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- tests/data/test_cve.py  3
- varats/data/provider/provider.py  2
- varats/plots/blame_interaction_degree.py  2
- varats/data/provider/cve/cve.py  8
- varats/data/provider/cve/cve_map.py  14
- varats/plots/cve_annotation.py  8
- varats/utils/git_util.py  1
- varats/tools/driver_cve.py  3
- varats/data/provider/cve/cve_provider.py  5
         

Clones added
============
- varats/projects/c_projects/lrzip.py  2
- varats/projects/c_projects/openssl.py  1
- varats/projects/c_projects/gzip.py  1
- varats/projects/c_projects/coreutils.py  1
- varats/projects/c_projects/tmux.py  2
- varats/projects/c_projects/glibc.py  1
- varats/projects/c_projects/vim.py  1
- varats/projects/c_projects/git.py  1
- varats/projects/c_projects/opus.py  2
- varats/projects/c_projects/libvpx.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5174691&bid=16985828)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- varats/plots/cve_annotation.py  8
- tests/data/test_cve.py  3
- varats/data/provider/provider.py  2
- varats/utils/git_util.py  1
- varats/plots/blame_interaction_degree.py  2
- varats/data/provider/cve/cve_map.py  14
- varats/tools/driver_cve.py  3
- varats/data/provider/cve/cve.py  8
- varats/data/provider/cve/cve_provider.py  5
         

Clones added
============
- varats/projects/c_projects/openssl.py  1
- varats/projects/c_projects/opus.py  2
- varats/projects/c_projects/gzip.py  1
- varats/projects/c_projects/tmux.py  2
- varats/projects/c_projects/coreutils.py  1
- varats/projects/c_projects/libvpx.py  1
- varats/projects/c_projects/git.py  1
- varats/projects/c_projects/vim.py  1
- varats/projects/c_projects/lrzip.py  2
- varats/projects/c_projects/glibc.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5174691&bid=16985828)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- varats/plots/blame_interaction_degree.py  2
- varats/data/provider/cve/cve_provider.py  5
- varats/data/provider/cve/cve_map.py  14
- tests/data/test_cve.py  3
- varats/utils/git_util.py  1
- varats/tools/driver_cve.py  3
- varats/plots/cve_annotation.py  8
- varats/data/provider/cve/cve.py  8
- varats/data/provider/provider.py  2
         

Clones added
============
- varats/projects/c_projects/vim.py  1
- varats/projects/c_projects/glibc.py  1
- varats/projects/c_projects/opus.py  2
- varats/projects/c_projects/tmux.py  2
- varats/projects/c_projects/gzip.py  1
- varats/projects/c_projects/openssl.py  1
- varats/projects/c_projects/coreutils.py  1
- varats/projects/c_projects/git.py  1
- varats/projects/c_projects/libvpx.py  1
- varats/projects/c_projects/lrzip.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5174691&bid=16985828)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/data/provider/provider.py  2
- varats/data/provider/cve/cve.py  8
- varats/tools/driver_cve.py  3
- varats/utils/git_util.py  1
- varats/plots/blame_interaction_degree.py  2
- varats/data/provider/cve/cve_provider.py  5
- tests/data/test_cve.py  3
- varats/data/provider/cve/cve_map.py  14
- varats/plots/cve_annotation.py  8
         

Clones added
============
- varats/projects/c_projects/tmux.py  2
- varats/projects/c_projects/libvpx.py  1
- varats/projects/c_projects/gzip.py  1
- varats/projects/c_projects/opus.py  2
- varats/projects/c_projects/vim.py  1
- varats/projects/c_projects/glibc.py  1
- varats/projects/c_projects/lrzip.py  2
- varats/projects/c_projects/coreutils.py  1
- varats/projects/c_projects/openssl.py  1
- varats/projects/c_projects/git.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5174691&bid=16985828)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/data/provider/provider.py  2
- varats/plots/blame_interaction_degree.py  2
- tests/data/test_cve.py  3
- varats/data/provider/cve/cve_provider.py  5
- varats/tools/driver_cve.py  3
- varats/data/provider/cve/cve_map.py  14
- varats/plots/cve_annotation.py  8
- varats/utils/git_util.py  1
- varats/data/provider/cve/cve.py  8
         

Clones added
============
- varats/projects/c_projects/libvpx.py  1
- varats/projects/c_projects/openssl.py  1
- varats/projects/c_projects/opus.py  2
- varats/projects/c_projects/tmux.py  2
- varats/projects/c_projects/coreutils.py  1
- varats/projects/c_projects/glibc.py  1
- varats/projects/c_projects/lrzip.py  2
- varats/projects/c_projects/gzip.py  1
- varats/projects/c_projects/vim.py  1
- varats/projects/c_projects/git.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5174691&bid=16985828)

File: docs/source/vara-ts-api/data_providers/cve_provider.rst [hunk 1]
```diff
@@ -0,0 +1,10 @@
+CVE Provider
+------------
+
+This provider allows access to CVEs associated with a project.
+It relies on a :class:`hook<varats.data.provider.cve.cve_provider.CVEProviderHook>` to know what CVEs belong to that project and uses heuristics to determine which revision fixed a certain CVE.
+
+.. automodule:: varats.data.provider.cve.cve_provider
+    :members:
+    :undoc-members:
+    :show-inheritance:
```
Comment (vulder) [inline docs/source/vara-ts-api/data_providers/cve_provider.rst:10]: should we add the other cve related modules/classes to this docs? It's kind hard to work with CVEs without docs but on the otherside I would not create special pages for everything

File: varats/data/provider/cve/cve_map.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+
+def __collect_via_commit_mgs(
+    commits: tp.List[str]
+) -> tp.Dict[str, tp.Dict[str, tp.Set[tp.Union[CVE, CWE]]]]:
+    """
+    Collect data about resolved CVE's/CWE's using the commit message.
+
+    Args:
+        commits: a list of commits
```
Comment (vulder) [inline varats/data/provider/cve/cve_map.py]:
```suggestion
        commits: a list of commits in textual form
```
as I understand it, commit means the full commit as a text?
should we also rename the arg name?

File: varats/data/provider/cve/cve_map.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+def __collect_via_version(
+    commits: tp.List[str], cve_list: tp.FrozenSet[CVE]
+) -> tp.Dict[str, tp.Dict[str, tp.Set[tp.Union[CVE, CWE]]]]:
+    """
+    Collect data about resolved CVE's using the tagged versions
+    and the vulnerable version list in the CVE's.
+
+    Args:
+        commits: a list of commits
```
Comment (vulder) [inline varats/data/provider/cve/cve_map.py]: same as above

File: varats/data/provider/cve/cve_map.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+
+def __collect_via_references(
+        commits: tp.List[str], cve_list: tp.FrozenSet[CVE], vendor: str,
+        product: str) -> tp.Dict[str, tp.Dict[str, tp.Set[tp.Union[CVE, CWE]]]]:
+    """
+    Collect data about resolved CVE' using the reference list in each CVE's.
+
+    Args:
+        commits: a list of commits
```
Comment (vulder) [inline varats/data/provider/cve/cve_map.py]: same

File: varats/data/provider/cve/cve_map.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+    return results
+
+
+def __merge_results(
+    result_list: tp.List[tp.Dict[str, tp.Dict[str, tp.Set[tp.Union[CVE, CWE]]]]]
+) -> tp.Dict[str, tp.Dict[str, tp.Set[tp.Union[CVE, CWE]]]]:
+    """
+    Merge a list of results into one dictionary.
+
```
Comment (vulder) [inline varats/data/provider/cve/cve_map.py:205]: `Args: ` missing, this should be a list of commit_hash cve mapping results, if I'm correct^^

File: varats/projects/c_projects/gzip.py [hunk 6]
```diff
@@ -77,6 +77,9 @@ def get_release_revisions(cls, release_type: ReleaseType) -> tp.List[str]:
...
             ]
         return [
-            h for h, tag in tagged_commits
-            if re.match(minor_release_regex, tag)
+            h for h, tag in tagged_commits if re.match(minor_release_regex, tag)
         ]
+
+    @classmethod
+    def get_cve_product_info(cls) -> tp.List[tp.Tuple[str, str]]:
+        return [("gzip", "gzip"), ("gnu", "gzip")]
```
Comment (vulder) [inline varats/projects/c_projects/gzip.py:85]: I think you mentioned to me that if one adds multiple vendor/product tuples they are all merged together, a I right?

File: varats/projects/c_projects/x264.py [hunk 7]
```diff
@@ -24,7 +25,7 @@
     shallow_clone=False,
     version_filter=project_filter_generator("x264"))
 class X264(Project):  # type: ignore
-    """ Video encoder x264 (fetched by Git) """
+    """Video encoder x264 (fetched by Git)"""
 
     NAME = 'x264'
     GROUP = 'c_projects'
```
Comment (vulder) [inline varats/projects/c_projects/x264.py:31]: there are no cves for x264?

**Review by vulder — CHANGES_REQUESTED** [review]

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/plots/blame_interaction_degree.py  2
- tests/data/test_cve.py  3
- varats/data/provider/cve/cve_provider.py  5
- varats/utils/git_util.py  1
- varats/plots/cve_annotation.py  8
- varats/data/provider/cve/cve.py  8
- varats/tools/driver_cve.py  3
- varats/data/provider/cve/cve_map.py  14
- varats/data/provider/provider.py  2
         

Clones added
============
- varats/projects/c_projects/gzip.py  1
- varats/projects/c_projects/coreutils.py  1
- varats/projects/c_projects/lrzip.py  2
- varats/projects/c_projects/openssl.py  1
- varats/projects/c_projects/opus.py  2
- varats/projects/c_projects/libvpx.py  1
- varats/projects/c_projects/vim.py  1
- varats/projects/c_projects/git.py  1
- varats/projects/c_projects/tmux.py  2
- varats/projects/c_projects/glibc.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5174691&bid=16985828)

File: varats/projects/c_projects/gzip.py (hunk 6 above)
Comment (boehmseb) [reply varats/projects/c_projects/gzip.py:85]: Yes, that's correct

File: varats/projects/c_projects/x264.py (hunk 7 above)
Comment (boehmseb) [reply varats/projects/c_projects/x264.py:31]: I could not find any.

File: docs/source/vara-ts-api/data_providers/cve_provider.rst (hunk 1 above)
Comment (boehmseb) [reply docs/source/vara-ts-api/data_providers/cve_provider.rst:10]: see 815ee35

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/plots/blame_interaction_degree.py  2
- varats/utils/git_util.py  1
- varats/data/provider/cve/cve.py  8
- varats/tools/driver_cve.py  3
- varats/data/provider/cve/cve_provider.py  5
- tests/data/test_cve.py  3
- varats/data/provider/provider.py  2
- varats/data/provider/cve/cve_map.py  14
- varats/plots/cve_annotation.py  8
         

Clones added
============
- varats/projects/c_projects/vim.py  1
- varats/projects/c_projects/gzip.py  1
- varats/projects/c_projects/glibc.py  1
- varats/projects/c_projects/git.py  1
- varats/projects/c_projects/opus.py  2
- varats/projects/c_projects/tmux.py  2
- varats/projects/c_projects/lrzip.py  2
- varats/projects/c_projects/coreutils.py  1
- varats/projects/c_projects/openssl.py  1
- varats/projects/c_projects/libvpx.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5174691&bid=16985828)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- tests/data/test_cve.py  3
- varats/data/provider/provider.py  2
- varats/data/provider/cve/cve.py  8
- varats/plots/blame_interaction_degree.py  2
- varats/data/provider/cve/cve_map.py  14
- varats/utils/git_util.py  1
- varats/data/provider/cve/cve_provider.py  5
- varats/tools/driver_cve.py  3
- varats/plots/cve_annotation.py  8
         

Clones added
============
- varats/projects/c_projects/tmux.py  2
- varats/projects/c_projects/vim.py  1
- varats/projects/c_projects/gzip.py  1
- varats/projects/c_projects/opus.py  2
- varats/projects/c_projects/coreutils.py  1
- varats/projects/c_projects/lrzip.py  2
- varats/projects/c_projects/libvpx.py  1
- varats/projects/c_projects/git.py  1
- varats/projects/c_projects/glibc.py  1
- varats/projects/c_projects/openssl.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5174691&bid=16985828)

**Review by vulder — APPROVED** [review]
lgtm

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- varats/tools/driver_cve.py  3
- varats/plots/blame_interaction_degree.py  2
- varats/data/provider/provider.py  2
- tests/data/test_cve.py  3
- varats/data/provider/cve/cve_provider.py  5
- varats/data/provider/cve/cve_map.py  14
- varats/utils/git_util.py  1
- varats/plots/cve_annotation.py  8
- varats/data/provider/cve/cve.py  8
         

Clones added
============
- varats/projects/c_projects/glibc.py  1
- varats/projects/c_projects/gzip.py  1
- varats/projects/c_projects/lrzip.py  2
- varats/projects/c_projects/coreutils.py  1
- varats/projects/c_projects/tmux.py  2
- varats/projects/c_projects/libvpx.py  1
- varats/projects/c_projects/openssl.py  1
- varats/projects/c_projects/git.py  1
- varats/projects/c_projects/vim.py  1
- varats/projects/c_projects/opus.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5174691&bid=16985828)

File: varats/data/provider/cve/cve_map.py [hunk 8]
```diff
@@ -239,13 +239,13 @@ def generate_cve_map(
...
-    
+
     Args:
         path: path to the git repo of the project to get the map for
         products: a list of tuples used for querying the CVE database
         end: newest revision to consider
         start: oldest revision to consider
-        only_precise: only include CVEs where an exact fixing commit can be 
+        only_precise: only include CVEs where an exact fixing commit can be
             identified
```
Comment (vulder) [inline varats/data/provider/cve/cve_map.py]:
```suggestion

```
you should remove all trailing whitespace :wink:

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/data/provider/cve/cve_map.py  14
- varats/utils/git_util.py  1
- varats/plots/blame_interaction_degree.py  2
- varats/data/provider/cve/cve_provider.py  5
- varats/tools/driver_cve.py  3
- varats/plots/cve_annotation.py  8
- varats/data/provider/provider.py  2
- varats/data/provider/cve/cve.py  8
- tests/data/test_cve.py  3
         

Clones added
============
- varats/projects/c_projects/opus.py  2
- varats/projects/c_projects/coreutils.py  1
- varats/projects/c_projects/glibc.py  1
- varats/projects/c_projects/libvpx.py  1
- varats/projects/c_projects/vim.py  1
- varats/projects/c_projects/gzip.py  1
- varats/projects/c_projects/git.py  1
- varats/projects/c_projects/tmux.py  2
- varats/projects/c_projects/lrzip.py  2
- varats/projects/c_projects/openssl.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5174691&bid=16985828)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/plots/cve_annotation.py  8
- varats/data/provider/cve/cve_map.py  14
- tests/data/test_cve.py  3
- varats/data/provider/provider.py  2
- varats/utils/git_util.py  1
- varats/data/provider/cve/cve.py  8
- varats/data/provider/cve/cve_provider.py  5
- varats/plots/blame_interaction_degree.py  2
- varats/tools/driver_cve.py  3
         

Clones added
============
- varats/projects/c_projects/openssl.py  1
- varats/projects/c_projects/opus.py  2
- varats/projects/c_projects/libvpx.py  1
- varats/projects/c_projects/gzip.py  1
- varats/projects/c_projects/tmux.py  2
- varats/projects/c_projects/coreutils.py  1
- varats/projects/c_projects/glibc.py  1
- varats/projects/c_projects/git.py  1
- varats/projects/c_projects/vim.py  1
- varats/projects/c_projects/lrzip.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5174691&bid=16985828)

**Pull request merged**